src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/StringContentAnalysis/StringContentAbstractValue.cs

```diff
@@ -0,0 +1,244 @@
+
+        public StringContentAbstractValue(string literal)
+        {
+            LiteralState = StringContainsState.Yes;
+            NonLiteralState = StringContainsState.No;
+            LiteralValues = ImmutableHashSet.Create(literal);
+            NonLiteralValues = ImmutableHashSet<IOperation>.Empty;
+        }
+
+        public StringContentAbstractValue(IOperation nonLiteral)
+        {
+            LiteralState = StringContainsState.No;
+            NonLiteralState = StringContainsState.Yes;
+            LiteralValues = ImmutableHashSet<string>.Empty;
+            NonLiteralValues = ImmutableHashSet.Create(nonLiteral);
+        }
+
+        private StringContentAbstractValue(StringContainsState literalState = StringContainsState.Maybe, StringContainsState nonLiteralState = StringContainsState.Maybe)
+            : this(literalState, nonLiteralState, ImmutableHashSet<string>.Empty, ImmutableHashSet<IOperation>.Empty)
+        {
+            Debug.Assert(literalState == StringContainsState.No || literalState == StringContainsState.Maybe);
+            Debug.Assert(nonLiteralState == StringContainsState.No || nonLiteralState == StringContainsState.Maybe);
+        }
+
+        private StringContentAbstractValue(StringContainsState literalState, StringContainsState nonLiteralState, ImmutableHashSet<string> literalValues, ImmutableHashSet<IOperation> nonLiteralValues)
+        {
+            LiteralState = literalState;
+            NonLiteralState = nonLiteralState;
+            LiteralValues = literalValues;
+            NonLiteralValues = nonLiteralValues;
+        }
+
+        /// <summary>
+        /// Indicates if this string variable contains string literals or not.
+        /// </summary>
+        public StringContainsState LiteralState { get; }
+
+        /// <summary>
+        /// Indicates if this string variable contains non literal string operands or not.
+        /// </summary>
+        public StringContainsState NonLiteralState { get; }
+
+        /// <summary>
+        /// Gets a collection of the string literals that could possibly make up the contents of this string <see cref="Operand"/>.
+        /// </summary>
+        public ImmutableHashSet<string> LiteralValues { get; }
+
+        /// <summary>
+        /// Gets a collection of the non literal string operations that could possibly make up the contents of this string <see cref="Operand"/>.
+        /// </summary>
+        public ImmutableHashSet<IOperation> NonLiteralValues { get; }
+        
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as StringContentAbstractValue);
+        }
+
+        public bool Equals(StringContentAbstractValue other)
+        {
+            return other != null &&
+                LiteralState == other.LiteralState &&
+                NonLiteralState == other.NonLiteralState &&
+                LiteralValues.Equals(other.LiteralValues) &&
```

Comment:
I don't think immutable collections do contents equality, they just use the default `object.Equals` I believe. Was that your intent here?